# example_code
Scripts that have helped with analysis  

**gc_content.py**  
Calculated gc content, gc count, N count, sequence length for all contigs and for contigs greater than 500b.p.  
  
**snpable_pos_3.py**  
From an alignment see which isolates have an N at a specific loci where other isolates collectively have 2 or more of the following bases: A,T,C,G.  
Useful for looking into alignments from snapperDB v2.  
  
**ggtree_vir_heatmap.R**  
Using R to add virulence data to better annotate a phylogenetic tree.
